### PR TITLE
Fix ArrayList error in Orchestration pipeline - 4.x

### PR DESCRIPTION
**Pull request by metmajer**
- Fix SerializationException
- Fix NULL in SSDS generation
- Fix SaaS bug in FinalizeStage

### CODE REVIEW

**Comment by metmajer** [comment]
@clemensutschig @s2oBCN I am currently not able to execute Unit tests locally. CodeNarc and Changelog need to be fixed as well.

**Comment by clemensutschig** [comment]
martin - this is buggy :(

/home/runner/work/ods-jenkins-shared-library/ods-jenkins-shared-library/src/org/ods/util/Util.groovy: 33: Apparent variable 'logger' was found in a static scope but doesn't refer to a local variable, static field or class. Possible causes:
You attempted to reference a variable in the binding or an instance variable from a static context.
You misspelled a classname or statically imported field. Please check the spelling.
You attempted to use a method 'logger' but left out brackets in a place not allowed by the grammar.
 @ line 33, column 17.
                   logger.debug ("Hit Jenkins serialization issue; attempt #: ${maxRetries - retries}")
                   ^

1 error

**Comment by clemensutschig** [comment]
ok I have fixed all the codenarc findings ... :) - the changelog I leave to you :)

**Comment by metmajer** [comment]
@s2oBCN please merge this branch into your PR https://github.com/opendevstack/ods-jenkins-shared-library/pull/755. @clemensutschig there still seem to be CodeNarc issues.

**Comment by s2oBCN** [comment]
> @s2oBCN please merge this branch into your PR #755. @clemensutschig there still seem to be CodeNarc issues.

Yes, it's what I was doing

**Comment by s2oBCN** [comment]
> @s2oBCN please merge this branch into your PR #755. @clemensutschig there still seem to be CodeNarc issues.

Can I refactor the code into 3 classes? Instead of only one utils:
- CollectionWrapper (or CollectionWithForloop)

A more specific class. That also if at the end something fails add logs and thow exception
- ShellWithRetry => it's easier to change all the shells calls
- ContextWithRetry

**Comment by metmajer** [comment]
@s2oBCN refactor if you see a need to

File: src/org/ods/util/CollectionWithForLoop.groovy [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package org.ods.util
+
+import com.cloudbees.groovy.cps.NonCPS
+
+class CollectionWithForLoop {
+
+    @NonCPS
```
Comment (metmajer) [inline src/org/ods/util/CollectionWithForLoop.groovy:7]: @s2oBCN please remove the NonCPS annotations, Closures might contain CPS code

**Comment by clemensutschig** [comment]
this is incomplete - at least for the finalize stage the changes are missing ...

**Review by clemensutschig — APPROVED** [review]

**Comment by clemensutschig** [comment]
please also fwd port to msater :)

**Comment by metmajer** [comment]
Thanks for the fixes @s2oBCN

**Pull request merged**